### ISSUE TITLE
cp.push_dir doesn't handle unicode filename or directory name

### ISSUE THREAD

**Issue by rajagopalx** [saltstack/salt#41018]
### Description of Issue

While pushing non english files and directory from minon, cp.push_dir returned False

#### Master:
    `salt '*' cp.push_dir D:/Files`

#### Minion:
   On Windows, I have Folder D:\Files which contains file, `திருக்குறள் Book.pdf`

#### Minion's Log

```
[salt.loaded.int.module.cp][DEBUG][3296] Trying to copy 'D:\Files\?????? Book.pdf' to master
[salt.loaded.int.module.cp][DEBUG][3296] Path failed os.path.isfile check, returning False
```
#### master

```
Salt Version:
           Salt: 2016.11.4
 
Dependency Versions:
           cffi: Not Installed
       cherrypy: Not Installed
       dateutil: 2.4.2
      docker-py: Not Installed
          gitdb: 0.6.4
      gitpython: 1.0.1
          ioflo: Not Installed
         Jinja2: 2.9.6
        libgit2: Not Installed
        libnacl: Not Installed
       M2Crypto: Not Installed
           Mako: 1.0.3
   msgpack-pure: Not Installed
 msgpack-python: 0.4.8
   mysql-python: Not Installed
      pycparser: Not Installed
       pycrypto: 2.6.1
   pycryptodome: Not Installed
         pygit2: Not Installed
         Python: 2.7.12 (default, Nov 19 2016, 06:48:10)
   python-gnupg: Not Installed
         PyYAML: 3.12
          PyZMQ: 15.2.0
           RAET: Not Installed
          smmap: 0.9.0
        timelib: Not Installed
        Tornado: 4.2.1
            ZMQ: 4.1.4
 
System Versions:
           dist: Ubuntu 16.04 xenial
        machine: x86_64
        release: 4.4.0-75-generic
         system: Linux
        version: Ubuntu 16.04 xenial
```

#### minion

`Windows 10 Salt: 2016.11.4`

**Comment by rajagopalx** [comment]
Here is the fix:

**https://github.com/saltstack/salt/blob/develop/salt/modules/cp.py#L782**

change:

```
for root, _, files in os.walk(path):

into

for root, _, files in os.walk(ur''.join(path)):
```

**Comment by gtmanfred** [comment]
Thanks for the PR for this.

Is that compatible with python2 and python3?

Thanks,
Daniel

**Comment by rajagopalx** [comment]
i ll check

**Comment by rajagopalx** [comment]
`ur''` is not compatible in python3.
We have to use 'r' instead of 'ur'

References:
http://bugs.python.org/issue15096
http://stackoverflow.com/questions/26063899/python-version-3-4-does-not-support-a-ur-prefix

**Comment by rajagopalx** [comment]
Also I added,

```
import sys
reload(sys)
sys.setdefaultencoding('utf-8')
```
in the cp.py file

**Comment by terminalmage** [comment]
I don't experience this same bug in Salt when the system locale is unicode. So the correct fix here is probably to ensure that the salt-minion service on Windows is started with a unicode locale.

**Comment by rajagopalx** [comment]
Is there any document on how to start service with unicode locale. i only found [this](https://support.microsoft.com/en-us/help/932659/how-to-use-the-new-language-locales-in-windows-server-2003-service-pack-2-sp2). Still though I am only to able choose specific language.

**Comment by terminalmage** [comment]
I'm not really familiar with locales in Windows. I have read that the unicode codepage (65001) has issues and is not recommended to be used by default. I made a commit in my fork (https://github.com/terminalmage/salt/commit/7ae83d32acac2b30b2f8b98911f2847bbde82e4b) which attempts to decode the strings obtained from the ``os.walk()``, but I just tested it in Windows and it didn't work.

Looping in @twangboy and @UtahDave who have more experience with dealing with Windows locales and may have better insight on the best way to proceed.
